### ISSUE TITLE
[no-unsafe-assignment] fails to infer the type from class properties

### ISSUE THREAD

**Issue by photonstorm** [typescript-eslint/typescript-eslint#2109]
**Repro**

```TS
import { IEventInstance } from './IEventInstance';

export class EventEmitter
{
    events: Map<string, Set<IEventInstance>>;

    constructor ()
    {
        this.events = new Map();
    }
}
```

**Expected Result**

The `no-unsafe-assignment` rule is flagging the creation of the Map in the constructor as being unsafe, claiming it can accept `any` types, even though the type is enforced by the `events` property in the class.

**Actual Result**

I see no logical reason why this should be considered an unsafe assignment when the type is clearly enforced by the property. TypeScript itself prevents incorrect insertion of values into this Map.

**Additional Info**

In order to stop the rule complaining, the following works:

```TS
import { IEventInstance } from './IEventInstance';

export class EventEmitter
{
    events: Map<string, Set<IEventInstance>>;

    constructor ()
    {
        this.events = new Map<string, Set<IEventInstance>>();
    }
}
```

However, this is just duplicating the type already defined.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.1` |
| `@typescript-eslint/parser`        | `3.0.1` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
Yeah I ran into this when attempting to introduce the rule into this codebase.

For some reason TS doesn't automatically infer the type of the RHS if it's `new Map()`.
This works 100% of the time for `new Set()`, so it's unique to map.

I _think_ it's due to how the `Map` constructor is defined.
https://github.com/microsoft/TypeScript/blob/6ac6a0410988e9f400c9153ac2b5fe9419849bf3/src/lib/es2015.collection.d.ts#L11-L16
```ts
interface MapConstructor {
    new(): Map<any, any>;
    new<K, V>(entries?: readonly (readonly [K, V])[] | null): Map<K, V>;
    readonly prototype: Map<any, any>;
}
declare var Map: MapConstructor;
```

Compared to `Set`:
https://github.com/microsoft/TypeScript/blob/6ac6a0410988e9f400c9153ac2b5fe9419849bf3/src/lib/es2015.collection.d.ts#L47-L51
```ts
interface SetConstructor {
    new <T = any>(values?: readonly T[] | null): Set<T>;
    readonly prototype: Set<any>;
}
declare var Set: SetConstructor;
```

Note that `Map` has an empty constructor which defaults to `any`s, and a parameterised constructor that doesn't.
OTOH `Set` only has a parameterised constructor, but its generic type has a default.

Unsure if it's a "bug" in TS or not. Either way, it's something that will have to be special cased in the rule unfortunately.

**Comment by bradzacher** [comment]
Doing some research, it looks like this is intentional.
Clicking through the [history for the lib def](https://github.com/microsoft/TypeScript/commits/master/src/lib/es2015.collection.d.ts), you find https://github.com/microsoft/TypeScript/pull/25374 which did _exactly what we want_!

However because `Map` has multiple type parameters, this caused weirdness with extending `Map` (https://github.com/microsoft/TypeScript/issues/27951), which then lead to just the change to just the `Map` constructor being reverted (https://github.com/microsoft/TypeScript/pull/28052).

So the map constructor being like this is completely intentional, which means we'll have to add a special case for `new Map()`.

**Comment by photonstorm** [comment]
Thanks for the quick update. Sounds like this is quite the rabbit hole! I've sadly found I have had to disable all of the `unsafe` rules for my codebase because they seem to conflict with TS in subtle ways. I don't envy you having to work with such a moving feast as this! Will keep an eye out at the Change Log.

**Comment by bradzacher** [comment]
> they seem to conflict with TS in subtle ways

How so? If there are problems with the rule, please report them so we can look into getting them fixed!

**Comment by Raynos** [comment]
I ran into this issues today as well. 

Would be lovely if `Map` can be special cased in `@typescript-eslint` because it's builtin into typescript stdlib definitions.

If I implement my own custom Generic `MapLike` and have an `any` signature in my own class definition then it's a "bug" in my constructor because I shouldn't use `any`.

Fixing TypeScript standard library for ES6 / etc to remove `any` is a large task ... so special casing `Map` in `@typescript-eslint` would be very practical.

**Comment by Raynos** [comment]
Here's an example workaround for JSDoc.

```js
class Foo {
  constructor () {
    // https://github.com/typescript-eslint/typescript-eslint/issues/1943
    /* eslint-disable @typescript-eslint/no-unsafe-assignment */
    /** @type {Map<string, FunctionConfiguration[]>} */
    this._profiles = /** @type {Map<string, FunctionConfiguration[]>} */ (new Map())
    /* eslint-enable @typescript-eslint/no-unsafe-assignment */
  }
}
```

JSdoc does not support `new Map<string, string>` but does support casts ... I opened an issue in TypeScript for the lacking JSDoc functionality in function calls ( https://github.com/microsoft/TypeScript/issues/38876 ).
